File: Hackerank/The Time in Words.cpp
#include <bits/stdc++.h>

using namespace std;

string ltrim(const string &);
string rtrim(const string &);

/*
 * Complete the 'timeInWords' function below.
 *
 * The function is expected to return a STRING.
 * The function accepts following parameters:
 *  1. INTEGER h
 *  2. INTEGER m
 */
string pasto(int x)
{
    if(x>=1 && x<=30)
        return "past";
    else
        return "to";
}

string timeInWords(int h, int m) {
    vector<string> numbers = {"zero",
        "one",
        "two",
        "three",
        "four",
        "five",
        "six",
        "seven",
        "eight",
        "nine",
        "ten",
        "eleven",
        "twelve",
        "thirteen",
        "fourteen",
        "fifteen",
        "sixteen",
        "seventeen",
        "eighteen",
        "nineteen",
        "twenty",
        "twenty one",
        "twenty two",
        "twenty three",
        "twenty four",
        "twenty five",
        "twenty six",
        "twenty seven",
        "twenty eight",
        "twenty nine"
        };
    string result;
    if(m==0)
    {
        result = numbers[h] + " o' clock";
    }
    else if(m==30)
    {
        result = "half past " + numbers[h];
    }
    else if(m==15)
    {
        result = "quarter past " + numbers[h];
    }
    else if(m==45)
    {
        result = "quarter to " + numbers[h+1];
    }
    else if(m==1)
    {
        result = numbers[m] + " minute past " + numbers[h];
    }
    else if(m>1 && m<=30)
    {
        result = numbers[m] + " minutes past " + numbers[h];
    }
    else
    {
        result = numbers[60-m] + " minutes to " + numbers[h+1];
    }
    return result;
    

}

int main()
{
    ofstream fout(getenv("OUTPUT_PATH"));

    string h_temp;
    getline(cin, h_temp);

    int h = stoi(ltrim(rtrim(h_temp)));

    string m_temp;
    getline(cin, m_temp);

    int m = stoi(ltrim(rtrim(m_temp)));

    string result = timeInWords(h, m);

    fout << result << "\n";

    fout.close();

    return 0;
}

string ltrim(const string &str) {
    string s(str);

    s.erase(
        s.begin(),
        find_if(s.begin(), s.end(), not1(ptr_fun<int, int>(isspace)))
    );

    return s;
}

string rtrim(const string &str) {
    string s(str);

    s.erase(
        find_if(s.rbegin(), s.rend(), not1(ptr_fun<int, int>(isspace))).base(),
        s.end()
    );

    return s;
}



/*

Sample Input:
5
47

Sample Output:
thirteen minutes to six

*/

File: Hackerank/Java/The time in words.java
import java.io.*;
import java.math.*;
import java.security.*;
import java.text.*;
import java.util.*;
import java.util.concurrent.*;
import java.util.regex.*;

public class Solution {

    // Complete the timeInWords function below.
    static String timeInWords(int h, int m) {
        String[] num = {"one","two","three","four","five","six","seven","eight","nine","ten","eleven","twelve","thirteen","fourteen","fifteen","sixteen","seventeen","eighteen","nineteen","twenty"};
        if (m == 00){
            return num[h-1]+" o' clock";
        }
        else{
            if (m < 30 ){
                if (m == 15){
                    return "quarter past "+num[h-1];
                }
                else{
                    if (m < 10){
                        return num[m-1]+" minute past "+num[h-1];
                    }
                    int temp = m % 10;
                    int val = m - temp;
                    if (val == 10){
                        return num[m-1]+ " minutes past "+num[h-1];
                    }
                    else{
                    return num[val-1] +" "+num[temp-1]+" minutes past "+num[h-1];
                    }
                }
            }
            else if (m == 30){
                return "half past "+num[h-1];
            }
            else{
                if (m == 45){
                    return "quarter to "+num[h];
                }
                else{
                    m = 60-m;
                    if (m < 10){
                        return num[m-1]+" minutes to "+num[h];
                    }
                    int temp = m % 10;
                    int val = m - temp;
                    if (val == 10){
                        return num[m-1]+ " minutes to "+num[h];
                    }
                    else{
                    return num[val-1] +" "+num[temp-1]+" minutes to "+num[h];
                    }
                }
            }
        }
        //return " ";
    }

    private static final Scanner scanner = new Scanner(System.in);

    public static void main(String[] args) throws IOException {
        BufferedWriter bufferedWriter = new BufferedWriter(new FileWriter(System.getenv("OUTPUT_PATH")));

        int h = scanner.nextInt();
        scanner.skip("(\r\n|[\n\r\u2028\u2029\u0085])?");

        int m = scanner.nextInt();
        scanner.skip("(\r\n|[\n\r\u2028\u2029\u0085])?");

        String result = timeInWords(h, m);

        bufferedWriter.write(result);
        bufferedWriter.newLine();

        bufferedWriter.close();

        scanner.close();
    }
}
